narwhal/types/src/primary.rs

```diff
@@ -1183,6 +1225,278 @@ impl CertificateV1 {
+                        votes.pop_front().unwrap();
+                    }
+                    return true;
+                }
+                false
+            })
+            .map(|(index, _)| index as u32);
+
+        let signed_authorities= roaring::RoaringBitmap::from_sorted_iter(filtered_votes)
+            .map_err(|_| DagError::InvalidBitmap("Failed to convert votes into a bitmap of authority keys. Something is likely very wrong...".to_string()))?;
+
+        // Ensure that all authorities in the set of votes are known
+        ensure!(
+            votes.is_empty(),
+            DagError::UnknownAuthority(votes.front().unwrap().0.to_string())
+        );
+
+        // Ensure that the authorities have enough weight
+        ensure!(
+            !check_stake || weight >= committee.quorum_threshold(),
+            DagError::CertificateRequiresQuorum
+        );
+
+        let aggregated_signature = if sigs.is_empty() {
+            AggregateSignature::default()
+        } else {
+            AggregateSignature::aggregate::<Signature, Vec<&Signature>>(
+                sigs.iter().map(|(_, sig)| sig).collect(),
+            )
+            .map_err(|_| DagError::InvalidSignature)?
+        };
+
+        let aggregate_signature_bytes = AggregateSignatureBytes::from(&aggregated_signature);
+
+        let aggregate_signature_state = if !check_stake {
+            AggregateSignatureState::Unsigned(aggregate_signature_bytes)
+        } else {
+            AggregateSignatureState::Unverified(aggregate_signature_bytes)
+        };
+
+        Ok(Certificate::V2(CertificateV2 {
+            header,
+            aggregate_signature_state,
+            signed_authorities,
+            metadata: Metadata::default(),
+        }))
+    }
+
+    /// This function requires that certificate was verified against given committee
+    pub fn signed_authorities(&self, committee: &Committee) -> Vec<PublicKey> {
+        assert_eq!(committee.epoch(), self.epoch());
+        let (_stake, pks) = self.signed_by(committee);
+        pks
+    }
+
+    pub fn signed_by(&self, committee: &Committee) -> (Stake, Vec<PublicKey>) {
+        // Ensure the certificate has a quorum.
+        let mut weight = 0;
+
+        let auth_indexes = self.signed_authorities.iter().collect::<Vec<_>>();
+        let mut auth_iter = 0;
+        let pks = committee
+            .authorities()
+            .enumerate()
+            .filter(|(i, authority)| match auth_indexes.get(auth_iter) {
+                Some(index) if *index == *i as u32 => {
+                    weight += authority.stake();
+                    auth_iter += 1;
+                    true
+                }
+                _ => false,
+            })
+            .map(|(_, authority)| authority.protocol_key().clone())
+            .collect();
+        (weight, pks)
+    }
+
+    /// Verifies the validity of the certificate.
+    /// TODO: Output a different type, similar to Sui VerifiedCertificate.
```

Comment:
@mwtian was thinking about removing this TODO as will have this distinction with `AggregateSignatureState`. Lmk what you think

Reply:
We can remove this TODO. I was hoping using an enum to represent signature and verification state would make migration easier. But there seem to be other issues. In particular, I think we still should not make `verify()` take a `&mut self`. How about:
- From `verify()` and also `sanitize()`, return a `Certificate` where the verification state is `VerifiedDirectly` for `CertificateV2`
- Add a `verify_indirectly()` function, that takes in a child certificate, checks the child is in a verified state and this certificate is one of its parent
?

Reply:
In this case, `verify()` and `sanitize()` will take `self` by value.

Reply:
Unfortunately it is very hard to migrate to returning a different `Certificate` type from `sanitize()` and `verify()`, but we can use a similar semantics for these functions. Usually functions named `sanitize()` and `verify()` should not mutate the underlying data.